rewriter/SelfNew.cc

```diff
@@ -0,0 +1,27 @@
+#include "rewriter/SelfNew.h"
+#include "ast/Helpers.h"
+#include "ast/ast.h"
+#include "core/core.h"
+
+namespace sorbet::rewriter {
+
+std::unique_ptr<ast::Expression> SelfNew::run(core::MutableContext ctx, ast::Send *send) {
+    if (send->fun != core::Names::new_() || !send->recv->isSelfReference()) {
```

Comment:
I think the existing tests should be enough to catch if this were to break, but it’s interesting to note now that there’s an order dependency among DSL passes now. For example:

- TEnum.cc special cases some logic for recognizing `X = new` as an enum definition
- Any pass that runs after this that wants to introduce `self.new` will have to remember to use Magic instead. 

Would you mind double checking point 2?

One way to make this a little less of a footgun is to create something like `ast::MK::new_` which would encapsulate the knowledge of making the Magic thing

Reply:
Another thing: @darkdimius do you know if there are implications wrt the name-based hashing things we do for LSP? Like after this intrinsic fires, it could appear to LSP like the NameRef `new` doesn’t appear in this AST anymore. Is that a problem?

Reply:
I think it is, but we don't have a test for it.
This is part of bigger problem that is true for many intrinsics(call with block, call with splat, etc)

Reply:
I don't think we should fix this in this PR, but this might be a good followup and shouldn't be hard(build a mapping for namerefs in the name ref collector)

Reply:
Gotcha; i've made a note to fix this. (trevor feel free to mark resolved if you want)